### PR TITLE
Added support to benchmark_model for cpu and mps

### PR DESCRIPTION
**Pull request by jainapurva**
Added support to benchmark_model for cpu and mps

Issue: https://github.com/pytorch/ao/issues/287

### CODE REVIEW

File: torchao/utils.py [hunk 1]
```diff
@@ -22,19 +24,49 @@
...
-    for _ in range(num_runs):
-        with torch.autograd.profiler.record_function("timed region"):
-            model(input_tensor)
-
-    end_event.record()
-    torch.cuda.synchronize()
-    return start_event.elapsed_time(end_event) / num_runs
+    if _assert_and_get_unique_device(model).type == "cuda":
+        torch.cuda.synchronize()
+        start_event = torch.cuda.Event(enable_timing=True)
```
Comment (msaroufim) [inline torchao/utils.py:29]: @albanD is there a reason why we have device specific events in PyTorch, it feels like this code would be considerably simpler if we just had a notion of an Event that prioritized cuda devices then mps then CPU

File: torchao/utils.py [hunk 2]
```diff
@@ -5,6 +5,8 @@
 from math import gcd
 from packaging import version
 import torch.nn.utils.parametrize as parametrize
+from torch.ao.quantization.utils import _assert_and_get_unique_device
```
Comment (msaroufim) [inline torchao/utils.py:8]: I can't seem to find this function on my end where is it defined?

```
(base) [marksaroufim@devgpu003.cco3 ~/ao (main)]$ grep -r "_assert_and_get_unique_device" .
```

Comment (jainapurva) [reply torchao/utils.py:8]: I'm using this function from https://github.com/pytorch/pytorch/blob/8a2fed7e6ab4cb6a97d92f6ca14f257370ec3a92/torch/ao/quantization/utils.py#L711

File: torchao/utils.py (hunk 1 above)
Comment (albanD) [reply torchao/utils.py:29]: If you're running latest PyTorch, do use `torch.Event` !

File: test/integration/test_integration.py [hunk 3]
```diff
@@ -1487,5 +1487,39 @@ def test_get_model_size_aqt(self, api, test_device, test_dtype):
...
+            x = self.linear2(x)
+            return x
+
+    def run_benchmark_model(self, device):
+        # params
+        dtype = torch.bfloat16
+        m = self.ToyLinearModel(1024, 1024, 1024).eval().to(dtype).to(device)
+        m_bf16 = copy.deepcopy(m)
+        example_inputs = m.example_inputs(dtype=dtype, device=device)
+        m_bf16 = torch.compile(m_bf16, mode='max-autotune')
```
Comment (msaroufim) [inline test/integration/test_integration.py:1512]: Number of runs should be closer to 1 and we might not need the max auto tune mode. The idea is here is this test should be as fast as possible so we just know that the benchmark function isn't broken. Intent is not to run a benchmark

**Review by msaroufim — CHANGES_REQUESTED** [review]

File: torchao/utils.py [hunk 4]
```diff
@@ -22,20 +24,71 @@
 ]
 
 
+def _assert_and_get_unique_device(module: torch.nn.Module) -> Any:
```
Comment (msaroufim) [inline torchao/utils.py:27]: Let's also make sure to remove this function from the quantization_utils file cause now code is just duplicated

**Review by msaroufim — APPROVED** [review]

File: torchao/utils.py (hunk 4 above)
Comment (jainapurva) [reply torchao/utils.py:28]: Function is referenced from https://github.com/pytorch/pytorch/blob/9105d54c6b37099575c0059ef274c86c4dc80c57/torch/ao/quantization/utils.py#L711

**Comment by jerryzh168** [comment]
> @pytorchbot merge --rebase

please use merge button for now, the bot is not ready yet

File: torchao/utils.py [hunk 5]
```diff
@@ -22,20 +24,72 @@
...
+    Returns the unique device for a module, or None if no device is found.
+    Throws an error if multiple devices are detected.
+    """
+    devices = {p.device for p in module.parameters()} | \
+        {p.device for p in module.buffers()}
+
+    if {torch.device("cpu"), torch.device("meta")} == devices:
+        warnings.warn("Both 'meta' and 'cpu' are present in the list of devices. Module can have one device. We Select 'cpu'.")
+        devices = {torch.device("cpu")}
+    ""
```
Comment (jerryzh168) [inline torchao/utils.py:39]: also remove this line

File: torchao/utils.py [hunk 6]
```diff
@@ -22,20 +24,72 @@
...
+        # benchmark
+        for _ in range(num_runs):
+            with torch.autograd.profiler.record_function("timed region"):
+                model(input_tensor)
+
+        end_event.record()
+        torch.cuda.synchronize()
+        return start_event.elapsed_time(end_event) / num_runs
+
+    elif _assert_and_get_unique_device(model).type == "mps":
```
Comment (jerryzh168) [inline torchao/utils.py:64]: nit: can you call the `_assert_and_get_unique_device` once?

**Pull request merged**